temporal-sdk/src/test/java/io/temporal/workflow/activityTests/LocalActivityRetryOverLocalThresholdTest.java

```diff
@@ -0,0 +1,168 @@
+
+    WorkflowException e =
+        assertThrows(
+            WorkflowException.class, () -> workflowStub.execute(testWorkflowRule.getTaskQueue()));
+    Assert.assertTrue(e.getCause() instanceof ActivityFailure);
+
+    controlledActivity.verifyAttempts();
+
+    List<HistoryEvent> historyEvents =
+        testWorkflowRule.getHistoryEvents(
+            WorkflowStub.fromTyped(workflowStub).getExecution(),
+            EventType.EVENT_TYPE_TIMER_STARTED);
+    assertEquals(
+        "12 attempts every 1 second with local activity retry threshold of 5 should take two timers",
+        2,
+        historyEvents.size());
+    historyEvents.forEach(
+        timerEvent ->
+            assertEquals(
+                ProtobufTimeUtils.toProtoDuration(Duration.ofSeconds(1)),
+                timerEvent.getTimerStartedEventAttributes().getStartToFireTimeout()));
+  }
+
+  @Test(timeout = 20_000)
+  public void maxAttemptDecreasedOnRetryWakeUp() {
+    Worker worker = testWorkflowRule.getWorker();
+    ControlledActivityImpl controlledActivity =
+        new ControlledActivityImpl(
+            Arrays.asList(
+                ControlledActivityImpl.Outcome.FAIL,
+                ControlledActivityImpl.Outcome.FAIL,
+                ControlledActivityImpl.Outcome.FAIL,
+                ControlledActivityImpl.Outcome.COMPLETE),
+            3,
+            -1);
+    worker.registerActivitiesImplementations(controlledActivity);
+    worker.registerWorkflowImplementationTypes(WorkflowReducingAttemptsImpl.class);
+    testWorkflowRule.getTestEnvironment().start();
+
+    TestWorkflows.TestWorkflow1 workflowStub =
+        testWorkflowRule.newWorkflowStubTimeoutOptions(TestWorkflows.TestWorkflow1.class);
+    WorkflowStub untypedStub = WorkflowStub.fromTyped(workflowStub);
+    WorkflowExecution execution = untypedStub.start(testWorkflowRule.getTaskQueue());
+
+    testWorkflowRule.waitForTheEndOfWFT(execution);
+    testWorkflowRule.invalidateWorkflowCache();
+
+    WorkflowException e =
+        assertThrows(WorkflowException.class, () -> untypedStub.getResult(String.class));
+    Assert.assertTrue(e.getCause() instanceof ActivityFailure);
+
+    controlledActivity.verifyAttempts();
+  }
+
+  public static class TestWorkflowImpl implements TestWorkflows.TestWorkflow1 {
+
+    @Override
+    public String execute(String taskQueue) {
+      LocalActivityOptions options =
+          LocalActivityOptions.newBuilder()
+              .setScheduleToCloseTimeout(Duration.ofSeconds(100))
+              .setLocalRetryThreshold(Duration.ofSeconds(5))
+              .setRetryOptions(
+                  RetryOptions.newBuilder()
+                      .setInitialInterval(Duration.ofSeconds(1))
+                      .setBackoffCoefficient(1)
+                      .setMaximumAttempts(12)
+                      .setDoNotRetry(AssertionError.class.getName())
+                      .build())
+              .build();
```

Comment:
We discussed on the call, but, for PR reference:

In core this wouldn't ever cause timer backoff to happen. That only happens if the amount we would back off exceeds the local retry threshold.

We agreed to adjust the meaning of local retry threshold to fit that concept. We'll also make it not user-adjustable, since it should really just be a function of workflow task timeout length.